### DISCUSSION TITLE
Trailing Stop Loss

### DISCUSSION THREAD

**Discussion by nigelkhakoo** [kernc/backtesting.py#238] [Q&A]
Hi - I am an experienced trader but am pretty new to coding /python.  I have implemented an event driving backtest using this library with self.buy() and self.sell().  Is there anyway to add trailing stop losses to this method of backtesting or do I have to use the vectorised backtesting method which I have not yet tried ?

Great library by the way.
Thanks

**Comment by kernc** [comment]
This is what [`TrailingStrategy`](https://kernc.github.io/backtesting.py/doc/backtesting/lib.html#backtesting.lib.TrailingStrategy) does under the hood:
https://github.com/kernc/backtesting.py/blob/a49122c72b8e0dd9b30104957638640653c2c113/backtesting/lib.py#L442-L450
Namely, it continually sets [trade's `.sl` properties](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Trade.sl).

These strategies in [`backtesting.lib`](https://kernc.github.io/backtesting.py/doc/backtesting/lib.html#header-classes) are meant to be independent and _composable_, meaning you might even be able to just inherit from them (as in [this tutorial](https://kernc.github.io/backtesting.py/doc/examples/Strategies%20Library.html)).

**Reply by nigelkhakoo** [reply]
OK got it - many thanks will try and code my own as i have a few different parameters

**Reply by kernc** [reply]
I'm curious, what other parameters do you have?

**Reply by nigelkhakoo** [reply]
Happy to discuss offline if thats ok

**Comment by atakanokan** [comment]
@kernc Since trailing SL is setting the `.sl` property at every bar, does this mean using SL and TSL are mutually exclusive? So it's impossible to set it up as SL @ 1% away from entry & TSL @ 2% away from entry like below?

```
    def next(self):
        # get previous bar close
        price = self.data["Close"][-1]

        # long/short SL and TP
        long_sl = price * 0.99
        long_tp = price * 1.01
        short_sl = price * 1.01
        short_tp = price * 0.99

        # signal logic
        if crossover(self.macd, self.macdsignal):
            self.buy(sl = long_sl, tp = long_tp)
        elif crossover(self.macdsignal, self.macd):
            self.sell(sl = short_sl, tp = short_tp)

        # tsl
        for trade in self.trades: 
            if trade.is_long: 
                trade.sl = max(trade.sl or -np.inf, 
                            self.data["Close"][-1] * 0.98) 
            else: # short
                trade.sl = min(trade.sl or np.inf, 
                            self.data["Close"][-1] * 1.02) 
```

**Reply by kernc** [reply]
Do you know of brokers where "trailing" is not just a modifier on the one and single SL order? Setting `sl=`/`.sl` and using `TrailingStrategy` _is_ mutually exclusive, but there is nothing preventing you from coding in an arbitrary SL logic.

To have initial SL at -1% and then trail with -2%, you could: 
```py
price = self.data.Close[-1]
if buy_condition:
    self.buy(sl=price*.99)
for trade in trades:
    if price / trade.entry_price > 1.01:     # Once profit is over >1%
        trade.sl = max(trade.sl, price*.98)  # "turn on" -2% trailing SL
```

**Reply by atakanokan** [reply]
Not aware of any brokers but it should be possible to code the strategy on client side to achieve the same behavior. And thanks for the example script really helpful!
